python/ray/ml/preprocessor.py

```diff
@@ -1,55 +1,101 @@
 import abc
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    import pandas as pd
 
 from ray.data import Dataset
 from ray.ml.predictor import DataBatchType
 
 
+class PreprocessorAlreadyFittedException(RuntimeError):
+    pass
+
+
+class PreprocessorNotFittedException(RuntimeError):
+    pass
+
+
 class Preprocessor(abc.ABC):
-    """Preprocessor interface"""
+    """Preprocessor interface for transforming Datasets."""
 
-    def fit_transform(self, ds: Dataset) -> Dataset:
-        """Fit this preprocessor to the given dataset.
+    # Preprocessors that do not need to be fitted must override this.
+    _is_fittable = True
 
-        Returns the fitted dataset.
+    def fit(self, dataset: Dataset) -> "Preprocessor":
+        """Fit this Preprocessor to the Dataset.
 
-        Args:
-            ds (Dataset): Input dataset.
+        Fitted state attributes will be directly set in the Preprocessor.
 
         Returns:
-            Dataset: Transformed dataset.
-
+            Preprocessor: The fitted Preprocessor with state attributes.
         """
-        raise NotImplementedError
+        if self.check_is_fitted():
+            raise PreprocessorAlreadyFittedException(
+                "`fit` cannot be called multiple times. "
+                "Create a new Preprocessor to fit a new Dataset."
+            )
 
-    def transform(self, ds: Dataset) -> Dataset:
-        """Transforms the given dataset.
+        return self._fit(dataset)
 
-        The preprocessor must be fitted first.
+    def _fit(self, dataset: Dataset) -> "Preprocessor":
+        raise NotImplementedError()
 
-        Args:
-            ds (Dataset): Input dataset.
+    def fit_transform(self, dataset: Dataset) -> Dataset:
+        """Fit this Preprocessor to the Dataset and then transform the Dataset.
 
+        Args:
+            dataset (Dataset): Input Dataset.
         Returns:
-            Dataset: Transformed dataset.
+            Dataset: The transformed Dataset.
+        """
+        self.fit(dataset)
+        return self.transform(dataset)
 
+    def transform(self, dataset: Dataset) -> Dataset:
+        """Transform the given dataset.
+
+        Args:
+            dataset (Dataset): Input Dataset.
+        Returns:
+            Dataset: The transformed Dataset.
         """
-        raise NotImplementedError
+        if self._is_fittable and not self.check_is_fitted():
+            raise PreprocessorNotFittedException(
+                "`fit` must be called before `transform`."
+            )
+        return self._transform(dataset)
+
+    def _transform(self, dataset: Dataset) -> Dataset:
+        return dataset.map_batches(self._transform_pandas, batch_format="pandas")
```

Comment:
Some or maybe even many of these preprocessors will want to provide/expose a `batch_size` for the batch transformation; the default is 4096 rows, but this might be too small for some preprocessors/datasets (bad performance) and too large for others (possible OOMs). For the latter, one example is a dataset holding a very wide table with thousands of columns, where. the default batch size might be too large to the point of causing OOMs.

Could we expose `batch_size` in the transformation API somehow? This could also wait for a follow-up PR.

Reply:
Ah great point. I wonder if this is the right API to expose or if there's something we can do to calculate the batch size internally. Adding a TODO for now!